File: QubitServer/src/org/labrad/qubits/channeldata/IqDataTime.java
package org.labrad.qubits.channeldata;

import java.util.concurrent.Future;

import org.labrad.qubits.channels.IqChannel;
import org.labrad.qubits.proxies.DeconvolutionProxy;
import org.labrad.qubits.proxies.DeconvolutionProxy.IqResult;
import org.labrad.qubits.util.ComplexArray;
import org.labrad.qubits.util.Futures;

import com.google.common.base.Function;
import com.google.common.base.Preconditions;

public class IqDataTime extends IqDataBase {

  private ComplexArray data;
  private int[] I, Q;

  public IqDataTime(ComplexArray data, boolean isDeconvolved) {
    this.data = data;
    if (isDeconvolved) {
      I = new int[data.re.length];
      Q = new int[data.im.length];
      for (int i = 0; i < data.length; i++) {
        I[i] = (int)(data.re[i] * 0x1fff) & 0x3fff;
        Q[i] = (int)(data.im[i] * 0x1fff) & 0x3fff;
      }
    }
    setDeconvolved(isDeconvolved);
  }

  public void checkLength(int expected) {
    LengthChecker.checkLengths(data.length, expected);
  }

  @Override
  public Future<Void> deconvolve(DeconvolutionProxy deconvolver) {
    IqChannel ch = getChannel();
    double freq = ch.getMicrowaveConfig().getFrequency();
    Future<DeconvolutionProxy.IqResult> req = deconvolver.deconvolveIq(ch.getDacBoard(), data, freq);
    return Futures.chain(req, new Function<DeconvolutionProxy.IqResult, Void>() {
      @Override
      public Void apply(IqResult result) {
        I = result.I;
        Q = result.Q;
        setDeconvolved(true);
        return null;
      }
    });
  }

  @Override
  public int[] getDeconvolvedI() {
    Preconditions.checkState(isDeconvolved(), "Data has not yet been deconvolved");
    return I;
  }

  @Override
  public int[] getDeconvolvedQ() {
    Preconditions.checkState(isDeconvolved(), "Data has not yet been deconvolved");
    return Q;
  }
}

File: QubitServer/src/main/scala/org/labrad/qubits/channeldata/IqDataTime.scala
package org.labrad.qubits.channeldata

import org.labrad.qubits.channels.IqChannel
import org.labrad.qubits.proxies.DeconvolutionProxy
import org.labrad.qubits.proxies.DeconvolutionProxy.IqResult
import org.labrad.qubits.util.ComplexArray
import scala.concurrent.{ExecutionContext, Future}

class IqDataTime(data: ComplexArray, isDeconvolved: Boolean, zeroEnds: Boolean) extends IqDataBase {

  @volatile private var I: Array[Int] = null
  @volatile private var Q: Array[Int] = null

  if (isDeconvolved) {
    I = data.re.map { i => (i * 0x1fff).toInt & 0x3fff }
    Q = data.im.map { q => (q * 0x1fff).toInt & 0x3fff }
  }
  setDeconvolved(isDeconvolved)

  def checkLength(expected: Int): Unit = {
    LengthChecker.checkLengths(data.length, expected)
  }

  override def deconvolve(deconvolver: DeconvolutionProxy)(implicit ec: ExecutionContext): Future[Unit] = {
    val ch = getChannel()
    val freq = ch.getMicrowaveConfig().frequency
    val req = deconvolver.deconvolveIq(ch.dacBoard, data, freq, zeroEnds)
    req.map { result =>
      I = result.I
      Q = result.Q
      setDeconvolved(true)
    }
  }

  override def getDeconvolvedI(): Array[Int] = {
    require(isDeconvolved(), "Data has not yet been deconvolved")
    I
  }

  override def getDeconvolvedQ(): Array[Int] = {
    require(isDeconvolved(), "Data has not yet been deconvolved")
    Q
  }
}

class IqDataTimeDacified(I: Array[Int], Q: Array[Int]) extends IqDataBase {

  setDeconvolved(true)

  def checkLength(expected: Int): Unit = {
    LengthChecker.checkLengths(I.length, expected)
    LengthChecker.checkLengths(Q.length, expected)
  }

  override def deconvolve(deconvolved: DeconvolutionProxy)(implicit ec: ExecutionContext): Future[Unit] = {
    sys.error("cannot deconvolve pre-dacified data")
  }

  override def getDeconvolvedI(): Array[Int] = I
  override def getDeconvolvedQ(): Array[Int] = Q
}
